### PR TITLE
Add Password reset View

### PR DESCRIPTION
**Pull request by silentninja**
Related to #1673 

PR adds a password reset view which the superuser would use to reset the password of any user. 

The API to call the Password Reset 
```
POST `/users/<user_id>/password_reset/`
{
  "password": str, //The new password of the user
}
```

The PR also adds a middleware to redirect users with default passwords to the password change page and the Django View for the password change page. 

Passwords are considered as `default`
- When a user's password is reset by the superuser

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by kgodey — CHANGES_REQUESTED** [review]
This approach won't work for the plan in the [Users and Permissions design issue](https://github.com/centerofci/mathesar/issues/1985). We'll be building the admin panel in Svelte (I assume) and the reset password functionality will need to be part of that panel, not a separate page. We'll need a reset password API to power that.

@pavish please confirm.

**Comment by pavish** [comment]
@silentninja @kgodey 

> This approach won't work for the plan in the https://github.com/centerofci/mathesar/issues/1985. We'll be building the admin panel in Svelte (I assume) and the reset password functionality will need to be part of that panel, not a separate page. We'll need a reset password API to power that.
> @pavish please confirm.

Yes, this is correct. We will be implementing all of the users and permissions frontend in svelte and will require APIs for them.

**Comment by silentninja** [comment]
@pavish Is there any reason to implement login APIs using svelte? 

There are a few reasons I wouldn't use Svelte
- Login  APIs don't hold state across pages which makes a tool like svelte unnecessary. I am not sure where Svelte would be used especially after your [comment](https://groups.google.com/a/mathesar.org/g/mathesar-developers/c/QSprKWKb5JQ/m/R-r4s1VuBAAJ) where you have mentioned using Django templates for the login APIs.
- The Django forms already have built-in features like validation and reimplementing these looks redundant.

File: mathesar/api/ui/viewsets/users.py [hunk 1]
```diff
@@ -16,6 +21,16 @@ class UserViewSet(AccessViewSetMixin, viewsets.ModelViewSet):
     pagination_class = DefaultLimitOffsetPagination
     access_policy = UserAccessPolicy
 
+    @action(methods=['post'], detail=False)
```
Comment (kgodey) [inline mathesar/api/ui/viewsets/users.py:24]: Why not make this a detail view? You don't need the username then.

File: mathesar/api/ui/permissions/users.py [hunk 2]
```diff
@@ -11,7 +11,7 @@ class UserAccessPolicy(AccessPolicy):
         },
         # Only superusers can create users
         {
-            'action': ['create'],
+            'action': ['create', 'password_reset'],
```
Comment (kgodey) [inline mathesar/api/ui/permissions/users.py:14]: Users should be able to reset their own passwords, it shouldn't just be superusers.

**Review by kgodey — CHANGES_REQUESTED** [review]

File: mathesar/api/ui/viewsets/users.py (hunk 1 above)
Comment (silentninja) [reply mathesar/api/ui/viewsets/users.py:24]: That makes sense if the front end is listing users and knows about the `id`. I was thinking of it as a form for the superuser to enter the username of the user. I will change it to a detail view

File: mathesar/api/ui/permissions/users.py (hunk 2 above)
Comment (kgodey) [reply mathesar/api/ui/permissions/users.py:14]: Okay yes, you're correct. We do need a reset password flow for users later (using email to verify), but that's not in scope right now.

Comment (silentninja) [reply mathesar/api/ui/permissions/users.py:14]: Created #2052 for the password change flow

**Review by kgodey — CHANGES_REQUESTED** [review]
@silentninja Sorry for not flagging this earlier, but the [users and permissions spec]() says:

> Administrators can reset anyone’s password. We should follow security best practices so that the admin does not ever see the user’s permanent password.

It seems like with this implementation, the admin does see the user's permanent password, since they are setting it. Can we avoid this? Either by forcing users to change passwords on first login (not sure how we'd do this with Svelte) or the reset password flow should generate a link that the admin can send to the user to set their own password (also not sure how we'd do this with Svelte). Please discuss with @pavish if needed.

**Comment by pavish** [comment]
@kgodey @silentninja 

I'm confused because the discussions here seem to be talking about different things. Here's what I have in mind regarding usage of Svelte/django templates:
1. The admin panel will be in Svelte. We need APIs for all operations here.
2. The login page will be in django templates. Only after successful login, we'll redirect to the app, which is in Svelte.
3. Password reset by users:
    - If the user wants to reset their own password because they can't login, it'll be part of the login page (django template). This is the [flow mentioned here](https://github.com/centerofci/mathesar/pull/2000#discussion_r1041446843).
4. Password change by users:
    - If the user wants to change their password, it'll be in Svelte. They can do this only after they're logged in already.
5. Password resets by admin:
    - This will be part of the admin panel, so everything will be in Svelte and be done using APIs.

Any UI after the user authentication is complete, will be in Svelte. Everything before that will be in Django templates.

## The password reset flow

### User account creation & first login
* I'm in favor of letting the admin type in the password, and then forcing the user to change their password at first login.
* The admin is the one creating the account, not the user. So the admin decides the username anyway, and it makes more sense for them to provide a password and force the user to change it.
* I don't see why we need a flow for a link for setting password here, since the admin has access to that link anyway, and that link is unsecure because anyone with that link can set the password for the target user.
* The implementation here would be in Django templates (no matter the flow), because the authentication process is not fully complete until the user changes their password and logs in again.
   - Secondly, we should not allow the user to access any other part of the app until they change their password.

### Password reset for existing user
* When the admin wants to reset a user's password, they should be allowed to type it in, and then force the user to change it at the next login, same as the previous case.
* Since the admin has full control of creating, deleting, and setting permissions for an account, this makes most sense.

The link for password reset seems like a good option only if the user themselves want to reset their password, because they've forgotten it, and we have a flow where the reset link is sent to their emails. I don't think it makes sense when the admin wants to reset the password.

**Comment by kgodey** [comment]
> The flow for forcing the user to set password will also be in Django templates.

This is the part I wasn't sure whether it should be in Django or Svelte, thanks for clarifying @pavish.

@silentninja we also need to implement this part, either in this PR or another PR. I think it makes sense to implement in this PR since it's related to this feature.

**Comment by silentninja** [comment]
> I don't see why we need a flow for a link for setting password here, since the admin has access to that link anyway

The same applies when an admin enters a password, he could log in with the same password and then change it. Our current method won't stop an admin if he is determined in seeing a users password.

**Comment by kgodey** [comment]
> The same applies when an admin enters a password, he could log in with the same password and then change it. Our current method won't stop an admin if he is determined to see the user's password.

We're not trying to stop admins from getting access to user accounts by any means possible. We're just trying to increase the default level of security.

**Comment by kgodey** [comment]
I've added a new "change password page" to the [Users & Permissions design checklist](https://github.com/centerofci/mathesar/issues/1985) based on the plan we've come up with here.

**Comment by silentninja** [comment]
@kgodey You might have to review the PR again as the commits after your last review adds a lot more features to the PR. I have updated the description to reflect the changes

**Comment by silentninja** [comment]
@kgodey @pavish When are we planning to update the styling for the Django templates?

**Comment by silentninja** [comment]
We should be considering passwords set by the superuser during user creation as default password too and redirect the user to change the password when they log in initially but I haven't made any changes to our existing user flow in this PR as it is out of scope for the current PR.

I will create a separate issue to track it.

**Comment by kgodey** [comment]
> When are we planning to update the styling for the Django templates?

Once @ghislaineguerin has completed #1985.

**Comment by kgodey** [comment]
> We should be considering passwords set by the superuser during user creation as default password too and redirect the user to change the password when they log in initially

Good catch, I added another checklist item in #1673 for it

**Pull request merged**